lfs_fs_preporphans: return int to alllow graceful LFS_ASSERT

mon opened the pull request:
From the discussion in #503, I started to implement the split `LFS_ASSERT`/`LFS_ASSERT_RETURN` functionality, but realised that all but one assert already exists in a function returning `int` or `lfs_ssize_t`.

Thus, by changing this function to return `int`, I can continue to use my custom LFS_ASSERT impl:
`#define LFS_ASSERT(test) do {if(!(test)) {log_error("lfs ASSERT %s:%d ", __FILE__, __LINE__); return LFS_ERR_ASSERT;}} while(0)`

Though there aren't any checks in place to prevent newly written LFS_ASSERT-using functions from returning `void`, I added `LFS_ERR_ASSERT` to the `lfs_error` enum to hopefully encourage consistency. Happy to revert that part of the commit.

geky commented:
Hi @mon, thanks for creating a PR.

Lets go with this. I've rejected previous proposals like this in the past, so hopefully no one else reading this is frustrated by the change in direction, but from the feedback this change has too much practical value.

So from this point on LFS_ASSERT will only be used in littlefs in functions that can return an int if users override the macro. Users are still free to use the macro in non-int functions, but they lose this ability. Though I guess they can probably just use their own assert macro in most cases.

> Though there aren't any checks in place to prevent newly written LFS_ASSERT-using functions from returning `void`

Ah! The way to handle this is a CI job that tests compilation with your error-returning assert. You could probably copy the readonly-compilation CI job and change the `-DLFS_READONLY` to `-D'LFS_ASSERT(test) do {if(!(test)) {log_error("lfs ASSERT %s:%d ", __FILE__, __LINE__); return LFS_ERR_ASSERT;}} while(0)'`. It's not the prettiest but apparently works ([according to this](https://stackoverflow.com/questions/31857559/gcc-define-function-like-macros-using-d-argument)):
https://github.com/littlefs-project/littlefs/blob/1a59954ec64ca168828a15242cc6de94ac75f9d1/.travis.yml#L211-L225

> I added `LFS_ERR_ASSERT` to the `lfs_error` enum to hopefully encourage consistency. Happy to revert that part of the commit.

Hmm, I've been been sticking to Linux errno codes since they offer a well-tested set of standard error codes that cover most system operations ([unix_system_error](https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html)). Unfortunately there isn't one for "ASSERT". I like using -1, which I've previously repurposed to a general "unknown" error code, though I've also seen EINVAL and EFAULT returned from assert failures.

I think it would be better to leave it out, since it does run the risk of confusing users on the default behavior of asserts. LFS_ERR_INVAL specifically may be preferred by users since littlefs also uses asserts to catch bad file configurations.

mon commented:
Alright! I've added that test, not using readonly since more assert-using code is built with the normal compile.

Have also reverted the new constant - I took EDOOFUS (Programming error) from BSD, but I think letting the user choose -1, EINVAL or similar works better.

geky commented:
>  took EDOOFUS (Programming error) from BSD

Oh that is a fun one! Unfortunately -88 is the less exciting ENOTSOCK (footwear is shoe) in Linux. For better or worse it also looks like some of the error codes are already Linux specific (ENOATTR looks different in BSD). It would have been a beautiful world if POSIX defined the actual integer values, but oh well.

---

Assuming CI passes, this looks perfect 👍. Thanks again for the PR and I will merge this into the next minor release.
